### PR TITLE
The resources change Request handler Migration

### PR DESCRIPTION
**Pull request by Anchellon**
## Summary

- Adds `POST /api/resources/{id}/change_requests` to the Go service, migrated from the Ruby `ResourceChangeRequest` handler
- Applies field changes immediately in a DB transaction (same approach as the existing phone change request handler)
- Stores a `pending` change request record and field-level audit entries in `change_requests` and `field_changes` tables

## Changes

**`internal/changerequest/types.go`**
Added `ResourceChangeRequestPayload`, `ResourceFields`, and `ResourceChangeRequest` types. A separate payload struct was needed because resources use a flat `change_request: { field: value }` shape. Fields are `*string` pointers so absent fields are `nil` and don't overwrite existing DB values.

**`internal/db/resources.go`**
Added `UpdateResource` — runs a transaction that updates the resources table, inserts a `change_requests` row, and inserts one `field_changes` row per changed field. Uses the existing `buildUpdateQuery` helper with an explicit allowlist of the 8 fields the frontend sends.

**`internal/changerequest/manager.go`**
Added `UpdateResource` HTTP handler. Also fixed two issues caught during testing: `fieldChangesResponse` initialized as `[]FieldChange{}` (not nil) so it marshals to `[]` not `null`; `writeStatus(201)` called before `writeJson()` so the status code is committed before the body is written.

**`cmd/sheltertech-go/main.go`**
Added route: `r.Post("/api/resources/{id}/change_requests", changeRequestManager.UpdateResource)`

## Design decisions

- **No approve/reject lifecycle** —  the Ruby codebase has an approve/reject flow but it's legacy debt. persist_change (which applies the update to the DB) is called on both create AND approve, meaning data is written twice. The Go migration applies the change immediately on creation, consistent with how the phone handler works. Approve/reject can be layered in later.
- **Allowlist over passthrough** — Ruby's resource change request has no permitted_fields restriction, accepting any column. The Go implementation uses an explicit allowlist of the 8 fields the frontend actually sends, keeping the surface area intentional and safe.

## Testing

Integration tests added to `resources_integration_test.go`:

- **SingleField** — happy path, asserts 201 + correct response shape + field in `field_changes`
- **EmptyBody** — `change_request: {}` returns 201 with empty `field_changes` array
- **InvalidID** — non-numeric resource ID returns 400
- **MultipleFields** — 3 fields sent, all 3 returned; uses `ElementsMatch` since map iteration order is non-deterministic

```bash
go test -v -tags=integration -run "TestCreateResourceChangeRequest" ./cmd/sheltertech-go/...
```

## The flow being mapped 

| Step | Ruby | Go |
|---|---|---|
| 1. Receive request | Rails router → `ChangeRequestsController#create` | `r.Post(...)` → `Manager.UpdateResource` |
| 2. Parse body | `params[:change_request]` | `json.Unmarshal` into `ResourceChangeRequestPayload` → `ResourceFields` |
| 3. Start transaction | ActiveRecord implicit | `tx.Begin()` |
| 4. Update resource row | `resource.update(field_change_hash)` | `tx.Exec(buildUpdateQuery(...))` |
| 5. Insert change request record | `ResourceChangeRequest.create(...)` | `tx.QueryRow(insertChangeRequestSql, "ResourceChangeRequest", ...)` |
| 6. Insert field change rows | saved via association | `tx.Exec(insertFieldChangeSql, ...)` per field |
| 7. Commit | ActiveRecord implicit | `tx.Commit()` |
| 8. Return response | Rails renders JSON | `writeStatus(201)` + `writeJson(response)` |

### CODE REVIEW

File: internal/changerequest/types.go [hunk 1]
```diff
@@ -33,3 +33,22 @@ type PhoneFields struct {
 	Number      *string `json:"number,omitempty"`
 	ServiceType *string `json:"service_type,omitempty"`
 }
+
+type ResourceChangeRequestPayload struct {
```
Comment (dfloo) [inline internal/changerequest/types.go]: We should use the generic ChangeRequestPayload from line 5 for all the various types. Maybe we should add 'omitempty' to the ParentResourceID and Type since they are only present on the create requests

File: internal/changerequest/manager.go [hunk 2]
```diff
@@ -35,6 +35,88 @@ func (m *Manager) Create(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (m *Manager) UpdateResource(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+
+	body, _ := io.ReadAll(r.Body)
+	payload := &ResourceChangeRequestPayload{}
+	err := json.Unmarshal(body, payload)
```
Comment (dfloo) [inline internal/changerequest/manager.go]: There's a helper function below  `unmarshalPayload` for the first stage of unmashaling

File: internal/changerequest/types.go (hunk 1 above)
Comment (Anchellon) [reply internal/changerequest/types.go]: The frontend sends fields flat inside change_request — no field_changes
wrapper, no action key (different from the phone change request handler):

```
{
  "change_request": 
  {
    "name": "...",
    "email": "...",
    "website": "...",
    "long_description": "...",
    "alternate_name": "...",
    "legal_status": "...",
    "internal_note": "..."
  }
}
```
Hence I have adopted a ResourceChangeRequestPayload

Comment (Anchellon) [reply internal/changerequest/types.go]: I do think that once we are completely off the ruby app , we should consider any updates to follow the json payload structure of something on the lines of 
```
{
  "type": "resources",
  "parent_resource_id": 1, // which will be nothing but the resource id itself
  "change_request": {
    "action": "update", /// for resources we will only have update as our current flow does resource creation without using the idea of a change request 
    "field_changes": { "email": "foo@bar.com", "long_description": "I am the description of an organization", .... other fields here }
 }

```

Keeping the type and resource id will be useful as it does leverage the switch case you have setup 


So to elaborate more on the Resources Entity specifically 

The creation as of now does not go through a change request endpoint 

It goes  through `POST /api/resources` which expects the 
The payload as follows
```{
  "resources": [{
    "name": "...",
    "addresses": [...],
    "long_description": "...",
    "email": "...",
    "website": "...",
    "notes": [{ "note": { "note": "..." } }],
    "schedule": { "schedule_days": [{ "day": "Monday", "opens_at": "09:00", "closes_at": "17:00" }] },
    "phones": [...],
    "legal_status": "..."
  }]
}
```
In a nutshell 
POST /api/resources → Rails creates the resource (with nested records) in one transaction, force-sets it to :approved, geocodes addresses, assigns SFSG site, returns the new ID → frontend reads the ID from the response and redirects to /organizations/{id}


Given that the payload being sent from the front end for this endpoint does not respect the generic ChangeRequestPayload I have proceeded with this.

Comment (Anchellon) [reply internal/changerequest/types.go]: Open to ideas @dfloo !

File: internal/changerequest/manager.go (hunk 2 above)
Comment (Anchellon) [reply internal/changerequest/manager.go]: I have updated it now after we got the frontend to update its request payload

**Comment by Anchellon** [comment]
Updated to use the generic change request payload after ensuring that the front end code is updated to match the json contract

**Review by dfloo — APPROVED** [review]

**Review by richardxia — APPROVED** [review]
LGTM!

I also just wanted to make a remark which is less about this PR, but more about the approach in general, which may be the best we can do based on the existing API and database schema. Because there are a lot more fields that can be edited on Resources compared to Phones, this PR is highlighting just how much boilerplate there is for each field. If we ever do a rewrite of a change approval flow, we should try to figure a way to do this that doesn't have so much boilerplate, even if that requires a completely different database schema.

**Review by richardxia — APPROVED** [review]
